app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala

```diff
@@ -612,48 +610,58 @@ class SectionRenderingService(frontendAppConfig: FrontendAppConfig, lookupRegist
   ): Html =
     renderUnit.fold {
       case RenderUnit.Pure(formComponent) =>
-        formComponent.`type` match {
-          case UkSortCode(_) =>
-            htmlForSortCode(formComponent, validationResult, ei)
-          case Group(_, _, _, _, _) =>
-            throw new IllegalArgumentException(s"Group '${formComponent.id}' cannot be rendered as RenderUnit.Pure")
-          case Date(_, offset, dateValue) =>
-            htmlForDate(formComponent, offset, dateValue, validationResult, ei)
-          case t @ Time(_, _) =>
-            renderTime(t, formComponent, validationResult, ei)
-          case Address(international) => htmlForAddress(formComponent, international, validationResult, ei)
-          case Text(Lookup(register, _), _, _, _, _, _) =>
-            renderLookup(formComponent, register, validationResult, ei)
-          case t @ Text(_, _, _, _, _, _) =>
-            renderText(t, formComponent, validationResult, ei)
-          case t @ TextArea(_, _, _, _, _) =>
-            renderTextArea(t, formComponent, validationResult, ei)
-          case Choice(choice, options, orientation, selections, hints, optionalHelpText) =>
-            htmlForChoice(
-              formComponent,
-              choice,
-              options,
-              orientation,
-              selections,
-              hints,
-              optionalHelpText,
-              validationResult,
-              ei)
-          case RevealingChoice(options, multiValue) =>
-            htmlForRevealingChoice(
-              formComponent,
-              multiValue,
-              formTemplateId,
-              options,
-              validationResult,
-              ei,
-              obligations)
-          case FileUpload() =>
-            htmlForFileUpload(formComponent, formTemplateId, ei, ei.retrievals, validationResult)
-          case InformationMessage(infoType, infoText) =>
-            htmlForInformationMessage(formComponent, infoType, infoText, ei)
-          case htp @ HmrcTaxPeriod(idType, idNumber, regimeType) =>
-            htmlForHmrcTaxPeriod(formComponent, ei, validationResult, obligations, htp)
+        val isVisible = formComponent.includeIf.fold(true) { includeIf =>
```

Comment:
why is the evaluation needed here? By the time we are here, the renderPageOptics would have already filtered invisible form components based on the includeIf evaluation right?

Reply:
No, `renderPageOptics` contains all the components. It is `formModelVisibilityOptics` which contain only visible ones.

The reason `renderPageOptics` contains all components, is that we need to know what components are on the page, so they get rendered accordingly.

Reply:
Got it, thanks